Allow BF programs to be read from a file.

a2aaron opened the pull request:
eval_file() takes a filepath and returns the evaluated memory array. It also supports input and output streams. Note that this function does NOT yet support reading input from files (although one could easily stream a file to it when run). I have not tested outputting to a file either.

Example programs have been added in a folder called "test_programs".

a2aaron committed:
Allow BF programs to be read from a file.

eval_file() takes a filepath and returns the evaluated memory array. It also supports input and output streams. Note that this function does NOT yet support reading input from files (although one could easilly stream a file to it when run). I have not tested outputing to a file either.

Example programs have been added in a folder called "test_programs".

a2aaron committed:
Merge branch 'develop' into feature/readfromfile.

Specifically, merge PR #7's extra tests into this branch.

porglezomp commented on test_brain_friendly.py:
```diff
@@ -1,8 +1,8 @@
 
 import pytest
-import io
+import io, sys
```
Don't import multiple modules on one line

a2aaron replied:
Fixed (Also, the "Bottles of Beer" test is actually reasonable now).

porglezomp commented:
Haha, when this gets merged you'll get credit for 1400+ lines of code added!

a2aaron committed:
Change "Bottles of Beer" test to be clearer.

The test now reads the example output as bytes and no longer requires version specific code.

Also, small style change.

porglezomp commented:
Change the Hello World test to use file IO

a2aaron committed:
Change "Hello World" test to read from I/O.

Also, tests for different varients of "Hello World" have been added.

porglezomp commented on brain_friendly.py:
```diff
@@ -85,6 +86,18 @@ def get_brace_matches(program):
 
+
+def eval_file(filename, index, memory, input=None, output=None):
+    f = open(filename, 'r')
```
change this to
```python
with open(filename, 'r') as f:
    program = f.read()
```

porglezomp commented on brain_friendly.py:
```diff
@@ -85,6 +86,17 @@ def get_brace_matches(program):
+def eval_file(filename, index, memory, input=None, output=None):
+    f = open(filename, 'r')
+    program = f.read()
+    f.close()
```
Change this to

``` python
with open(filename, 'r') as f:
    program = f.read()
```

since it automatically handles closing the file, and is prettier.

porglezomp commented on brain_friendly.py:
```diff
@@ -85,6 +86,17 @@ def get_brace_matches(program):
+    output = io.BytesIO()
+    bf_memory = eval_file(filename, bf_index, bf_memory, output=output)
+    output.seek(0)
+    print(output.getvalue())
```
There's no need to mess around with an `io.BytesIO` object, that means you won't see any output until the program finishes running! Instead, write directly to the `stdout` file stream. Since `sys.stdout` in Python 3 is a Unicode stream, I recommend either

``` python
if hasattr(sys.stdout, 'buffer'):
    output = sys.stdout.buffer
else:
    output = sys.stdout
```

or, probably better, since it's EAFP,

``` python
try:
    output = sys.stdout.buffer
except AttributeError:
    output = sys.stdout
```

Both of these will write directly to the output stream, which will make it more responsive, and also conceptually simpler.

porglezomp commented on brain_friendly.py:
```diff
@@ -85,6 +86,17 @@ def get_brace_matches(program):
-    print(bf_memory[:10])
+    filename = "test_programs/hello_world/hello_world_commented.bf"
+    output = io.BytesIO()
+    bf_memory = eval_file(filename, bf_index, bf_memory, output=output)
```
You probably don't need to assign to `bf_memory`, since we're not using it.

porglezomp commented on test_brain_friendly.py:
```diff
@@ -186,63 +187,43 @@ def test_invalid_braces():
+    file_outputs = [b'Hello World!\n', b'Hello World!\n',
+                    b'Hello World!\n', b'hello world', b'Hello, World!']
+
+    for (filename, file_output) in zip(filenames, file_outputs):
```
You don't need parenthesis around `(filename, file_output)`.

porglezomp commented on test_brain_friendly.py:
```diff
@@ -186,63 +187,43 @@ def test_invalid_braces():
->++.                    And finally a newline from Cell #6'''
+
+
+# Test reading from a file
```
These all seem to be tests of more interesting programs. We should have some dead simple programs that let us test slightly more robustly, so if we break something in the interpreter, some of the file IO tests will still pass.

porglezomp commented on test_brain_friendly.py:
```diff
@@ -186,63 +187,43 @@ def test_invalid_braces():
+    '_no_comma2.bf', '_lowercase.bf', '_no_newline.bf']
+    filenames = [prefix+suffix for suffix in suffixes]
+    # Not all of these are the same.
+    file_outputs = [b'Hello World!\n', b'Hello World!\n',
```
I feel icky about storing the outputs in a separate list from the file names.
Perhaps better:

``` python
import os  # Do this at the top

prefix = 'test_programs/hello_world'
test_cases = [
    ('commented.bf', b'Hello World!\n'),
    ('no_comma.bf', b'Hello World!\n'),
    # ...
]
for file_name, file_output in test_cases:
    file_path = os.path.join(prefix, file_name)
    # Test with the file_path and file_output...
```

and then rename files like `hello_world/hello_world_commented.bf` to `hello_world/commented.bf`, since the directory makes the second `hello_world` redundant.

a2aaron committed:
Make Hello World tests clearer and cleaner looking.

Rename some test programs as well.

a2aaron committed:
Make brain_friendly.py output to standard output.

a2aaron committed:
Add a few simple file I/O tests.

This is to test just file I/O and not much else.

A tiny change to brain_friendly to have it use `with` when opening files (no actual logic change).
